DOC: Added MultiIndex Example for Series Min

vadakattu opened the pull request:
Corollary to #23298
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Commit:
DOC: Added multi-indexed example for series.max

Commit:
DOC: truncated line for PEP compliance

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-max

Commit:
DOC: Added Examples for Series Min

Commit:
DOC: Typo corrected.

Commit:
DOC: Restructured indexing as requested.

Commit:
DOC: Switched from MultiIndex.from_arrays to MultiIndex.from_product

Commit:
Merge branch 'master' into series-min

# Conflicts:
#	pandas/core/generic.py

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

# Conflicts:
#	pandas/core/generic.py

Commit:
Formatting Indentation for PEP compliance

jreback commented:
cc @datapythonista @WillAyd

datapythonista reviewed (approved):
lgtm, but we should avoid repeating this for every stats method, I'd prefer a template than repeating these examples many times.

jreback commented:
@vadakattu can you template this as @datapythonista suggests?

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

Commit:
DOC: Initial templating of stat_func multiIndex examples

datapythonista reviewed (commented):
Unfortunately we need to keep compatibility for Python 2.7 and 3.5 for some time more, so we can't use f-strings yet.

vadakattu commented:
Of course! ...that was very absent minded of me.

Do you have any suggestions for a better way to do the templating? Otherwise I'll go ahead and switch it to old-style formatting.

datapythonista commented:
@vadakattu can you please merge master into your branch and push, and make sure the CI is green, so we can merge this. Thanks!

Commit:
DOC: Refactored templating of stat_func multiIndex examples for Python 2.7

Commit:
DOC: Removed spurious f

datapythonista commented on pandas/core/generic.py:
```diff
@@ -10266,41 +10293,16 @@ def _doc_parms(cls):
+--------
+%(multi_index_example)s
+""" % dict(multi_index_example=_shared_docs['stat_func_multi_index'] %
+           _shared_docs['min_multi_index'])
```
Can't you have the `Examples` header in `_shared_docs['stat_func_multi_index']` and avoid all these complex an repetitive blocks of code?

Can you also render one of the docstring with `validate_docstrings.py` and copy the output in a comment. It's difficult to see what is finally being rendered with all this complexity.

vadakattu replied:
```
Return the sum of the values for the requested axis

Parameters
----------
axis : {index (0)}
skipna : boolean, default True
    Exclude NA/null values when computing the result.
level : int or level name, default None
    If the axis is a MultiIndex (hierarchical), count along a
    particular level, collapsing into a scalar
numeric_only : boolean, default None
    Include only float, int, boolean columns. If None, will attempt to use
    everything, then use only numeric data. Not implemented for Series.
min_count : int, default 0
    The required number of valid values to perform the operation. If fewer than
    ``min_count`` non-NA values are present the result will be NA.

    .. versionadded :: 0.22.0

       Added with the default being 0. This means the sum of an all-NA
       or empty Series is 0, and the product of an all-NA or empty
       Series is 1.

Returns
-------
sum : scalar or Series (if level specified)

Examples
--------
``MultiIndex`` series example of monthly rainfall

>>> index = pd.MultiIndex.from_product(
...     [['London', 'New York'], ['Jun', 'Jul', 'Aug']],
...     names=['city', 'month'])
>>> s = pd.Series([47, 35, 54, 112, 117, 113], index=index)
city      month
London    Jun       47
          Jul       35
          Aug       54
New York  Jun      112
          Jul      117
          Aug      113
dtype: int64

>>> s.sum()
478

Sum using level names, as well as indices.

>>> s.sum(level='city')
city
London      136
New York    342
dtype: int64

>>> s.sum(level=1)
month
Jun    159
Jul    152
Aug    167
dtype: int64

By default, the sum of an empty or all-NA Series is ``0``.

>>> pd.Series([]).sum()  # min_count=0 is the default
0.0

This can be controlled with the ``min_count`` parameter. For example, if
you'd like the sum of an empty series to be NaN, pass ``min_count=1``.

>>> pd.Series([]).sum(min_count=1)
nan

Thanks to the ``skipna`` parameter, ``min_count`` handles all-NA and
empty series identically.

>>> pd.Series([np.nan]).sum()
0.0

>>> pd.Series([np.nan]).sum(min_count=1)
nan
`

datapythonista reviewed (changes requested):
Added a comment, apply to all cases.

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

Commit:
DOC: Refactoring for simplicity

Commit:
DOC: Corrected None to empty string

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

Commit:
DOC: Formatting

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

vadakattu commented:
Validate doc strings seems to have a bug with the example validation which is causing failure for one of the build jobs -- it's not picking up the lines immediately beneath it.  (Applies to all 3 functions)

Full output of validate doc string below:

```
################################################################################
######################## Docstring (pandas.Series.min)  ########################
################################################################################

This method returns the minimum of the values in the object.
            If you want the *index* of the minimum, use ``idxmin``. This is
            the equivalent of the ``numpy.ndarray`` method ``argmin``.

Parameters
----------
axis : {index (0)}
skipna : boolean, default True
    Exclude NA/null values when computing the result.
level : int or level name, default None
    If the axis is a MultiIndex (hierarchical), count along a
    particular level, collapsing into a scalar
numeric_only : boolean, default None
    Include only float, int, boolean columns. If None, will attempt to use
    everything, then use only numeric data. Not implemented for Series.

Returns
-------
min : scalar or Series (if level specified)

Examples
--------
``MultiIndex`` series example of monthly rainfall

>>> index = pd.MultiIndex.from_product(
...     [['London', 'New York'], ['Jun', 'Jul', 'Aug']],
...     names=['city', 'month'])
>>> s = pd.Series([47, 35, 54, 112, 117, 113], index=index)
city      month
London    Jun       47
          Jul       35
          Aug       54
New York  Jun      112
          Jul      117
          Aug      113
dtype: int64

>>> s.min()
35

Minimize using level names, as well as indices.

>>> s.min(level='city')
city
London       35
New York    112
dtype: int64

>>> s.min(level=1)
month
Jun    47
Jul    35
Aug    54
dtype: int64

################################################################################
################################## Validation ##################################
################################################################################

11 Errors found:
        Docstring text (summary) should start in the line immediately after the opening quotes (not in the same line, or leaving a blank line in between)
        Use only one blank line to separate sections or paragraphs
        Summary must start with infinitive verb, not third person (e.g. use "Generate" instead of "Generates")
        Summary should fit in a single line.
        Errors in parameters section
                Parameters {**kwargs} not documented
                Parameter "axis" has no description
                Parameter "skipna" type should use "bool" instead of "boolean"
                Parameter "level" description should finish with "."
                Parameter "numeric_only" type should use "bool" instead of "boolean"
        Examples do not pass tests
1 Warnings found:
        See Also section not found

################################################################################
################################### Doctests ###################################
################################################################################

**********************************************************************
Line 30, in pandas.Series.min
Failed example:
    s = pd.Series([47, 35, 54, 112, 117, 113], index=index)
Expected:
    city      month
    London    Jun       47
              Jul       35
              Aug       54
    New York  Jun      112
              Jul      117
              Aug      113
    dtype: int64
Got nothing
```

I can also looking into correcting the pre-existing issues with this functions parameter documentation if you'd like. Else I can leave that for a separate PR.

datapythonista commented:
The validation seems right.

```
>>> s = pd.Series([47, 35, 54, 112, 117, 113], index=index)
city      month
London    Jun       47
          Jul       35
          Aug       54
New York  Jun      112
          Jul      117
          Aug      113
dtype: int64
```

Is wrong. If you assign the `Series` to `s`, nothing is returned by this. Are you missing a `>>> s` in the next line?

vadakattu commented:
Got what you mean -- I think the slight bug / unexpected behaviour is the wording of the message

Sounds like it should be:
```
Got: ...
Expected: nothing
```
instead of the other way around.

Commit:
DOC: Corrected example output

datapythonista commented on pandas/core/generic.py:
```diff
@@ -10416,44 +10438,54 @@ def _doc_parms(cls):
+                                       additional='')
+
+_sum_examples = _shared_docs['stat_func_example'] % \
+    _shared_docs['sum_multi_index']
```
I think it's better if we do:
```
_sum_examples = _shared_docs['stat_func_example'].format(
    stat_func='sum', verb='Sum', default_ouput='478',
    ...)
```

That will makes things simpler, having some `_shared_docs` with strings, and some with dicts, was quite confusing to me, took a decent amount of time to understand.

I think it'll probably make things simpler to use `'...{stat_func}...'.format(stat_func='sum')` than the one with a `%` in this case (but if it doesn't make things simpler, you can leave the old one.

I'd prefer an example where it's faster to understand the results. For example, if I show you `op(2, 3) == 5` it's quite easy to see that the `op` is summing. But with higher magnitudes, like the `op(112, 47)` it takes more time to make the math and make sure you're understanding the examples correctly.

We've been using in many places examples with animals and their number of legs, and things like that. I think it's probably a good idea to use them here.

For the `additional` variable, as it comes after the examples, I'd simply add them at the end like:
```
_sum_examples = _shared_docs['stat_func_example'].format(
    ....)
_sum_examples += '''
the rest of the example
'''
```
I think this makes things much easier to read and find too.

vadakattu replied:
.format() with keyword arguments is not allowed by old-style formatting. I propose the following as the closest approximation to what you want:
```
_sum_examples = _shared_docs['stat_func_example'] % dict(
    stat_func='sum',
    verb='Sum',
    ...)
```

jreback commented:
@datapythonista how does this look

datapythonista commented:
@vadakattu do you have time to update the PR based on the previous comment?

vadakattu commented:
Will do.

datapythonista commented:
@vadakattu do you have time to address that last comment?

vadakattu commented:
yes, will do this week.

Commit:
DOC: Updated stat func base example

Commit:
DOC: Updated stat func sum, min and max examples for clarity.

Commit:
DOC: Updated string variables from old format -> new.

vadakattu commented:
Done.

datapythonista commented on pandas/core/generic.py:
```diff
@@ -10397,43 +10397,50 @@ def _doc_parms(cls):
 """
 
-_sum_examples = """\
+_shared_docs['stat_func_example'] = """\
```
is the backslash needed?

datapythonista commented on pandas/core/generic.py:
```diff
@@ -10397,43 +10397,50 @@ def _doc_parms(cls):
+...     ['Warm', 'Warm', 'Cold', 'Cold'],
+...     ['Dog', 'Falcon', 'Fish', 'Spider']],
+...     names=['Blooded', 'Animal'])
+>>> s = pd.Series([4, 2, 0, 8], name='Legs', index=index)
```
I think in the rest of the documentation we use the values, columns... (e.g. dog, warm, blooded...) in lower case. Can you take a quick look to confirm, and change it please.

And we use `idx` for the index.

vadakattu replied:
Done

datapythonista commented on pandas/core/generic.py:
```diff
@@ -10397,43 +10397,50 @@ def _doc_parms(cls):
 Examples
 --------
-``MultiIndex`` series example of monthly rainfall
+``MultiIndex`` series example.
```
If this is the only example, I'd remove this sentence, as it's not adding much value. If we left it, MultiIndex would go between single backticks, and Series with capital S (I know it's from the original docstring, just letting you know).

datapythonista reviewed (commented):
Thanks @vadakattu, Looks good, couple of small things to make this standard with the rest of the docs.

Also, if you can run `./scripts/validate_docstrings.py pandas.Series.min`... and check that there is nothing wrong with the docstrings based on the script, that would be great.

datapythonista commented:
I think this is modifying some the same things as #22554. @vadakattu @Roald87, can you coordinate and avoid duplicate work

Commit:
DOC: Formatting update.

Commit:
DOC: Updated sum, min & max for complete doc string compliance.

vadakattu commented:
Have now restructured `_num_doc`, and incorporated the minimal changes necessary across the stat functions (as desired in #22554) such min, max & sum have fully compliant docstrings.

`./scripts/validate_docstrings.py pandas.Series.min`
```
################################################################################
######################## Docstring (pandas.Series.min)  ########################
################################################################################

Return the minimum of the values for the requested axis.

            If you want the *index* of the minimum, use ``idxmin``. This is
            the equivalent of the ``numpy.ndarray`` method ``argmin``.

Parameters
----------
axis : {index (0)}
    Axis for the function to be applied on.
skipna : bool, default True
    Exclude NA/null values when computing the result.
level : int or level name, default None
    If the axis is a MultiIndex (hierarchical), count along a
    particular level, collapsing into a scalar.
numeric_only : bool, default None
    Include only float, int, boolean columns. If None, will attempt to use
    everything, then use only numeric data. Not implemented for Series.
**kwargs
    Additional keyword arguments to be passed to the function.

Returns
-------
min : scalar or Series (if level specified)

Examples
--------

>>> idx = pd.MultiIndex.from_arrays([
...     ['warm', 'warm', 'cold', 'cold'],
...     ['dog', 'falcon', 'fish', 'spider']],
...     names=['blooded', 'animal'])
>>> s = pd.Series([4, 2, 0, 8], name='legs', index=idx)
>>> s
blooded  animal
warm     dog       4
         falcon    2
cold     fish      0
         spider    8
Name: legs, dtype: int64

>>> s.min()
0

Min using level names, as well as indices.

>>> s.min(level='blooded')
blooded
warm    2
cold    0
Name: legs, dtype: int64

>>> s.min(level=0)
blooded
warm    2
cold    0
Name: legs, dtype: int64

See Also
--------
Series.sum : Return the sum.
Series.min : Return the minimum.
Series.max : Return the maximum.
Series.idxmin : Return the index of the minimum.
Series.idxmax : Return the index of the maximum.
DataFrame.min : Return the sum over the requested axis.
DataFrame.min : Return the minimum over the requested axis.
DataFrame.max : Return the maximum over the requested axis.
DataFrame.idxmin : Return the index of the minimum over the requested axis.
DataFrame.idxmax : Return the index of the maximum over the requested axis.

################################################################################
################################## Validation ##################################
################################################################################

Docstring for "pandas.Series.min" correct. :)
```
`./scripts/validate_docstrings.py pandas.Series.max`
```
################################################################################
######################## Docstring (pandas.Series.max)  ########################
################################################################################

Return the maximum of the values for the requested axis.

            If you want the *index* of the maximum, use ``idxmax``. This is
            the equivalent of the ``numpy.ndarray`` method ``argmax``.

Parameters
----------
axis : {index (0)}
    Axis for the function to be applied on.
skipna : bool, default True
    Exclude NA/null values when computing the result.
level : int or level name, default None
    If the axis is a MultiIndex (hierarchical), count along a
    particular level, collapsing into a scalar.
numeric_only : bool, default None
    Include only float, int, boolean columns. If None, will attempt to use
    everything, then use only numeric data. Not implemented for Series.
**kwargs
    Additional keyword arguments to be passed to the function.

Returns
-------
max : scalar or Series (if level specified)

Examples
--------

>>> idx = pd.MultiIndex.from_arrays([
...     ['warm', 'warm', 'cold', 'cold'],
...     ['dog', 'falcon', 'fish', 'spider']],
...     names=['blooded', 'animal'])
>>> s = pd.Series([4, 2, 0, 8], name='legs', index=idx)
>>> s
blooded  animal
warm     dog       4
         falcon    2
cold     fish      0
         spider    8
Name: legs, dtype: int64

>>> s.max()
8

Max using level names, as well as indices.

>>> s.max(level='blooded')
blooded
warm    4
cold    8
Name: legs, dtype: int64

>>> s.max(level=0)
blooded
warm    4
cold    8
Name: legs, dtype: int64

See Also
--------
Series.sum : Return the sum.
Series.min : Return the minimum.
Series.max : Return the maximum.
Series.idxmin : Return the index of the minimum.
Series.idxmax : Return the index of the maximum.
DataFrame.min : Return the sum over the requested axis.
DataFrame.min : Return the minimum over the requested axis.
DataFrame.max : Return the maximum over the requested axis.
DataFrame.idxmin : Return the index of the minimum over the requested axis.
DataFrame.idxmax : Return the index of the maximum over the requested axis.

################################################################################
################################## Validation ##################################
################################################################################

Docstring for "pandas.Series.max" correct. :)
```
`./scripts/validate_docstrings.py pandas.Series.sum`
```
################################################################################
######################## Docstring (pandas.Series.sum)  ########################
################################################################################

Return the sum of the values for the requested axis.

            This is equivalent to the method ``numpy.sum``.

Parameters
----------
axis : {index (0)}
    Axis for the function to be applied on.
skipna : bool, default True
    Exclude NA/null values when computing the result.
level : int or level name, default None
    If the axis is a MultiIndex (hierarchical), count along a
    particular level, collapsing into a scalar.
numeric_only : bool, default None
    Include only float, int, boolean columns. If None, will attempt to use
    everything, then use only numeric data. Not implemented for Series.
min_count : int, default 0
    The required number of valid values to perform the operation. If fewer than
    ``min_count`` non-NA values are present the result will be NA.

    .. versionadded :: 0.22.0

       Added with the default being 0. This means the sum of an all-NA
       or empty Series is 0, and the product of an all-NA or empty
       Series is 1.
**kwargs
    Additional keyword arguments to be passed to the function.

Returns
-------
sum : scalar or Series (if level specified)

Examples
--------

>>> idx = pd.MultiIndex.from_arrays([
...     ['warm', 'warm', 'cold', 'cold'],
...     ['dog', 'falcon', 'fish', 'spider']],
...     names=['blooded', 'animal'])
>>> s = pd.Series([4, 2, 0, 8], name='legs', index=idx)
>>> s
blooded  animal
warm     dog       4
         falcon    2
cold     fish      0
         spider    8
Name: legs, dtype: int64

>>> s.sum()
14

Sum using level names, as well as indices.

>>> s.sum(level='blooded')
blooded
warm    6
cold    8
Name: legs, dtype: int64

>>> s.sum(level=0)
blooded
warm    6
cold    8
Name: legs, dtype: int64

By default, the sum of an empty or all-NA Series is ``0``.

>>> pd.Series([]).sum()  # min_count=0 is the default
0.0

This can be controlled with the ``min_count`` parameter. For example, if
you'd like the sum of an empty series to be NaN, pass ``min_count=1``.

>>> pd.Series([]).sum(min_count=1)
nan

Thanks to the ``skipna`` parameter, ``min_count`` handles all-NA and
empty series identically.

>>> pd.Series([np.nan]).sum()
0.0

>>> pd.Series([np.nan]).sum(min_count=1)
nan

See Also
--------
Series.sum : Return the sum.
Series.min : Return the minimum.
Series.max : Return the maximum.
Series.idxmin : Return the index of the minimum.
Series.idxmax : Return the index of the maximum.
DataFrame.min : Return the sum over the requested axis.
DataFrame.min : Return the minimum over the requested axis.
DataFrame.max : Return the maximum over the requested axis.
DataFrame.idxmin : Return the index of the minimum over the requested axis.
DataFrame.idxmax : Return the index of the maximum over the requested axis.

################################################################################
################################## Validation ##################################
################################################################################

Docstring for "pandas.Series.sum" correct. :)
```

datapythonista commented:
@vadakattu can you fix the conflicts? we merged the PR that I mentioned it was touching same code as this one

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

# Conflicts:
#	pandas/core/generic.py

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

# Conflicts:
#	pandas/core/generic.py

vadakattu commented:
I believed I dealt with it last night -- the single circle ci failure seems unrelated.
Pulling master again to re-test.

```
=================================== FAILURES ===================================
___________________________ test_tick_add_sub[Milli] ___________________________

cls = <class 'pandas.tseries.offsets.Milli'>

    @pytest.mark.parametrize('cls', tick_classes)
>   @example(n=2, m=3)
    @example(n=800, m=300)
    @example(n=1000, m=5)
    @given(n=st.integers(-999, 999), m=st.integers(-999, 999))
    def test_tick_add_sub(cls, n, m):

pandas/tests/tseries/offsets/test_ticks.py:42: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/conda/envs/pandas-dev/lib/python3.6/site-packages/hypothesis/core.py:552: in execute
    ) % (test.__name__, text_repr[0],))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <hypothesis.core.StateForActualGivenExecution object at 0x7fb7828ac5f8>
message = 'Hypothesis test_tick_add_sub(cls=Milli, n=0, m=100) produces unreliable results: Falsified on the first call but did not on a subsequent one'

    def __flaky(self, message):
        if len(self.falsifying_examples) <= 1:
>           raise Flaky(message)
E           hypothesis.errors.Flaky: Hypothesis test_tick_add_sub(cls=Milli, n=0, m=100) produces unreliable results: Falsified on the first call but did not on a subsequent one

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

vadakattu commented:
Tests pass, and docstrings fully compliant per `validate_docstrings`.

Roald87 commented:
@datapythonista seems as if @vadakattu already made the full docs, so there's no point in merging my PR I guess?

datapythonista commented:
I didn't check in detail how much overlap exists in both PRs. But as pointed out in https://github.com/pandas-dev/pandas/pull/23338#issuecomment-443470861, you should talk with @vadakattu and see if it makes sense to have all the changes in one and close the other, or what.

Roald87 commented:
@vadakattu what do you think? I'm fine with taking yours, I don't mine adds much.

vadakattu commented:
👍🏽  Sounds good.

datapythonista commented:
@vadakattu can you fix the conflicts please. Sorry for the delay.

Commit:
Merge branch 'master' into series-min

# Conflicts:
#	pandas/core/generic.py

Commit:
Merge branch 'master' of github.com:pandas-dev/pandas into series-min

# Conflicts:
#	pandas/core/generic.py

vadakattu commented:
Conflicts resolved.

datapythonista reviewed (approved):
lgtm, thanks @vadakattu

WillAyd commented:
Thanks @vadakattu
